stdlib/immutablemap.gr

```diff
@@ -0,0 +1,495 @@
+/**
+ * Determines if the map contains the given key. In such a case, it will always contain a value for the given key.
+ *
+ * @param key: The key to search for
+ * @param map: The map to search
+ * @returns `true` if the map contains the given key or `false` otherwise
+ *
+ * @since v0.5.4
+ */
+export let rec contains = (key, map) => {
+  Option.isSome(get(key, map))
+}
+
+// removes the minimum element from a tree
+let rec removeMin = node => {
+  match (node) {
+    Tree({ left: Empty, right, _ }) => right,
+    Tree({ key, val, left, right, _ }) =>
+      balancedNode(key, val, removeMin(left), right),
+    _ => fail "Impossible: ImmutableMap removeMin on empty node",
+  }
+}
+
+// helper function for removing a node by creating a new node containing the
+// removed node's left and right subtrees
+let removeInner = (left, right) => {
+  match ((left, right)) {
+    (Empty, node) | (node, Empty) => node,
+    (left, right) => {
+      let (minKey, minVal) = min(right)
+      balancedNode(minKey, minVal, left, removeMin(right))
+    },
+  }
+}
+
+/**
+ * Produces a new map without the key-value pair corresponding to the given
+ * key. If the key doesn't exist in the map, the map will be returned unmodified.
+ *
+ * @param key: The key to exclude
+ * @param map: The map to exclude from
+ * @returns A new map without the given key
+ *
+ * @since v0.5.4
+ */
+export let rec remove = (key, map) => {
+  match (map) {
+    Empty => Empty,
+    Tree({ key: nodeKey, val: nodeVal, left, right, _ }) => {
+      match (compare(key, nodeKey)) {
+        cmp when cmp < 0 =>
+          balancedNode(nodeKey, nodeVal, remove(key, left), right),
+        cmp when cmp > 0 =>
+          balancedNode(nodeKey, nodeVal, left, remove(key, right)),
+        _ => removeInner(left, right),
+      }
+    },
+  }
+}
+
+/**
+ * Produces a new map by calling an updater function that receives the
+ * previously stored value as an `Option` and returns the new value to be
+ * stored as an `Option`. If the key didn't exist previously, the new value
+ * will be `None`. If `None` is returned from the updater function, the
+ * key-value pair is excluded.
+ *
+ * @param key: The unique key in the map
+ * @param fn: The updater function
+ * @param map: The map to modify
+ * @returns A new map with the value at the given key modified according to the function's output
+ *
+ * @since v0.5.4
+ */
+export let update = (key, fn, map) => {
+  let val = get(key, map)
+  match (fn(val)) {
+    Some(next) => set(key, next, map),
+    None => remove(key, map),
+  }
+}
+
+/**
+ * Iterates the map, calling an iterator function with each key and value.
+ *
+ * @param fn: The iterator function to call with each key and value
+ * @param map: The map to iterate
+ *
+ * @since v0.5.4
+ */
+export let forEach = (fn, map) => {
+  let rec forEachInner = node => {
+    match (node) {
+      Empty => void,
+      Tree({ key, val, left, right, _ }) => {
+        forEachInner(left)
+        fn(key, val)
```

Comment:
```suggestion
        fn(key, val): Void
```

As with other `forEach` in the stdlib, this needs to be annotated with `Void` otherwise users can return any value. I caught this by looking over the generated markdown